src/objective-c/tests/PerfTests/PerfTests.m

```diff
@@ -265,37 +262,40 @@ - (void)testPingPongRPCWithV1API {
+                     request:(RMTSimpleRequest *)request
+                 numMessages:(int)numMessages
+              numOutstanding:(int)numOutstanding
+                 callOptions:(GRPCMutableCallOptions *)options {
   __weak XCTestExpectation *expectation = [self expectationWithDescription:@"unaryRPC"];
 
-  dispatch_semaphore_t sema = dispatch_semaphore_create(kOutstandingRPCs);
+  dispatch_semaphore_t sema = dispatch_semaphore_create(numOutstanding);
   __block int index = 0;
 
-  for (int i = 0; i < numMessages; ++i) {
-    GRPCUnaryProtoCall *call = [_service
-        unaryCallWithMessage:request
-             responseHandler:[[PerfTestsBlockCallbacks alloc]
-                                 initWithInitialMetadataCallback:nil
-                                                 messageCallback:nil
-                                                   closeCallback:^(NSDictionary *trailingMetadata,
-                                                                   NSError *error) {
-                                                     dispatch_semaphore_signal(sema);
-                                                     @synchronized(self) {
-                                                       ++index;
-                                                       if (index == numMessages) {
-                                                         [expectation fulfill];
+  for (RMTTestService *service in services) {
+    for (int i = 0; i < numMessages; ++i) {
+      GRPCUnaryProtoCall *call = [service
+          unaryCallWithMessage:request
+               responseHandler:[[PerfTestsBlockCallbacks alloc]
+                                   initWithInitialMetadataCallback:nil
+                                                   messageCallback:nil
+                                                     closeCallback:^(NSDictionary *trailingMetadata,
+                                                                     NSError *error) {
+                                                       dispatch_semaphore_signal(sema);
+                                                       @synchronized(self) {
+                                                         ++index;
+                                                         if (index ==
+                                                             numMessages * [services count]) {
+                                                           [expectation fulfill];
+                                                         }
                                                        }
-                                                     }
 
-                                                   }]
-                 callOptions:options];
+                                                     }]
+                   callOptions:options];
 
-    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
-    [call start];
+      dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
```

Comment:
Waiting forever is usually not the best thing to do since if it got stuck in CI due to some bug, the process takes looooong time to timeout. I would try to avoid using forever but use some reasonably large number.

Reply:
Why do we control the number of outstanding calls?

Reply:
Fixed wait forever -> wait until timeout. 
Regarding controlling number of outstanding calls: this is something we do in some of our existing benchmarks (see go/grpc-benchmark-comparison-dashboard). I have tried to follow the same pattern here.